builtin/mainmenu/dlg_create_world.lua

```diff
@@ -51,46 +154,172 @@ local function create_world_formspec(dialogdata)
 				end
 			end
 		end
+
+		local ds = (gameconfig:get("disallowed_settings") or ""):split()
+		for _, value in pairs(ds) do
+			disallowed_settings[value:trim()] = true
+		end
 	end
 
 	local mglist = ""
-	local selindex = 1
+	local selindex
 	local i = 1
+	local first_mg
 	for k,v in pairs(mapgens) do
+		if not first_mg then
+			first_mg = v
+		end
 		if current_mg == v then
 			selindex = i
 		end
 		i = i + 1
 		mglist = mglist .. v .. ","
 	end
+	if not selindex then
+		selindex = 1
+		current_mg = first_mg
+	end
 	mglist = mglist:sub(1, -2)
 
+	local mg_main_flags = function(mapgen, y)
+		if mapgen == "singlenode" then
+			return "", y
+		end
+		if disallowed_settings["mg_flags"] then
+			return "", y
+		end
+
+		local form = "checkbox[4.2," .. y .. ";flag_mg_caves;" ..
+			fgettext("Caves") .. ";"..chflag(flags.main, "caves").."]"
+		y = y + 0.5
+		form = form .. "checkbox[4.2,"..y..";flag_mg_dungeons;" ..
+			fgettext("Dungeons") .. ";"..chflag(flags.main, "dungeons").."]"
+		y = y + 0.5
+
+		if mapgen == "v6" then
+			form = form .. "checkbox[4.2,"..y..";flag_mg_decorations;" ..
+			fgettext("Decorations (without trees and jungle grass)") .. ";"
+			.. chflag(flags.main, "decorations").."]" ..
+			"tooltip[flag_mg_decorations;" ..
+			fgettext("Decorations are special structures provided by the game that appear on the surface. Note: This does not affect trees and jungle grass generated by the v6 mapgen")
+			.. "]"
+
+			y = y + 0.5
+		else
+			form = form .. "checkbox[4.2,"..y..";flag_mg_decorations;" ..
+			fgettext("Decorations") .. ";" ..
+			chflag(flags.main, "decorations").."]" ..
+			"tooltip[flag_mg_decorations;" ..
+			fgettext("Decorations are special structures that generate on the surface includes things such as plants, trees and other small extras")
+			.. "]"
+
+			y = y + 0.5
+		end
+		form = form .. "tooltip[flag_mg_caves;" ..
+		fgettext("Carve a large, interconnected network of tunnels and caves in the terrain")
+		.. "]"
+		return form, y
+	end
+
+	local mg_specific_flags = function(mapgen, y)
+		if flag_checkboxes[mapgen] then
+			if disallowed_settings["mg"..mapgen.."_spflags"] then return "" end
+			local form = ""
+			for _,tab in pairs(flag_checkboxes[mapgen]) do
+				local id = "flag_mg"..mapgen.."_"..tab[1]
+				form = form .. "checkbox[4.2," .. y .. ";" .. id .. ";" .. tab[2] .. ";" ..
+					chflag(flags[mapgen], tab[3]).."]"
+				if tab[4] then
+					form = form .. "tooltip["..tab[1]..";"..tab[4].."]"
+				end
+				y = y + 0.5
+			end
+
+			if mapgen == "v6" then
```

Comment:
For reasons explained in a line comment above, i disapprove of adding so much perfectionist complexity just for v6. Users can read documentation to see how the biome flags work if they need to.
We only keep V6 around to support old worlds. V6 is at the end of the mapgen list to discourage use, new worlds are rarely started using v6. So this amount of complex specialisation for it is not justified.

Which biomes result from the flags is somewhat obvious already, so there is little justification for this special treatment. It seems to be for tooltips to show an exact list of biomes, which is unnecessary.

So please have v6 taken care of by the simple non-mgv6 code.

Reply:
This is fine to me if all mvg6 information is together in one table. Comment below.

Please return here like so to reduce indents:
```Lua
if mapgen ~= "v6" then
	-- No special treatment
	return "", y
end
```